File: resources/patterns/design/twin-pattern/twin-widget.ts
/**
 * Twin Pattern — runnable TypeScript example.
 *
 * A `Widget` keeps the shared state and the public API. `Graphic` and
 * `Interactive` are the twins: each owns one concern and holds a
 * back-reference to the widget so it can read shared state.
 *
 * Run: npx tsx twin-widget.ts   (or compile with tsc and run node)
 */

/** Minimal surface a twin needs from its widget — injectable for tests. */
export interface WidgetLike {
  name: string;
  x: number;
  y: number;
  width: number;
  height: number;
}

function linked(w: WidgetLike | null, twin: string): WidgetLike {
  if (!w) throw new Error(`${twin} twin is not linked to a widget`);
  return w;
}

/** Twin A: drawing behavior. */
export class Graphic {
  widget: WidgetLike | null = null;

  draw(): string {
    const w = linked(this.widget, 'Graphic');
    return `Drawing ${w.name} at (${w.x}, ${w.y})`;
  }

  resize(width: number, height: number): string {
    const w = linked(this.widget, 'Graphic');
    w.width = width;
    w.height = height;
    return `Resized to ${width}x${height}`;
  }
}

/** Twin B: interaction behavior. */
export class Interactive {
  widget: WidgetLike | null = null;

  onClick(): string {
    return `Clicked on ${linked(this.widget, 'Interactive').name}`;
  }

  onHover(): string {
    return `Hovering over ${linked(this.widget, 'Interactive').name}`;
  }
}

/** The composite twin class that links Graphic and Interactive. */
export class Widget implements WidgetLike {
  width = 100;
  height = 50;

  private graphic = new Graphic();
  private interactive = new Interactive();

  constructor(
    public name: string,
    public x = 0,
    public y = 0,
  ) {
    this.graphic.widget = this;
    this.interactive.widget = this;
  }

  draw(): string {
    return this.graphic.draw();
  }

  resize(width: number, height: number): string {
    return this.graphic.resize(width, height);
  }

  onClick(): string {
    return this.interactive.onClick();
  }

  onHover(): string {
    return this.interactive.onHover();
  }

  getGraphic(): Graphic {
    return this.graphic;
  }

  getInteractive(): Interactive {
    return this.interactive;
  }
}

const isMain = process.argv[1]?.endsWith('twin-widget.ts')
  || process.argv[1]?.endsWith('twin-widget.js');

if (isMain) {
  const button = new Widget('SubmitButton', 10, 20);
  console.log(button.draw());
  console.log(button.onClick());
  console.log(button.onHover());
  console.log(button.resize(200, 60));
}

File: resources/patterns/design/twin-pattern/twin_widget.py
"""Twin Pattern — runnable example.

A ``Widget`` keeps the shared state and the public API. ``Graphic`` and
``Interactive`` are the twins: each owns one concern and holds a
back-reference to the widget so it can read shared state.

Run the demo:

    python twin_widget.py

Run the tests:

    python -m unittest twin_widget -v
"""

from __future__ import annotations

import unittest
from typing import Optional, Protocol


class WidgetLike(Protocol):
    """Minimal surface a twin needs from its widget.

    Injecting this interface (instead of the concrete Widget) is what
    makes twins unit-testable: a stub with the same fields is enough.
    """

    name: str
    x: int
    y: int
    width: int
    height: int


class Graphic:
    """Twin A: drawing behavior."""

    def __init__(self) -> None:
        self.widget: Optional[WidgetLike] = None

    def draw(self) -> str:
        w = self._linked()
        return f"Drawing {w.name} at ({w.x}, {w.y})"

    def resize(self, width: int, height: int) -> str:
        w = self._linked()
        w.width = width
        w.height = height
        return f"Resized to {width}x{height}"

    def _linked(self) -> WidgetLike:
        if self.widget is None:
            raise RuntimeError("Graphic twin is not linked to a widget")
        return self.widget


class Interactive:
    """Twin B: interaction behavior."""

    def __init__(self) -> None:
        self.widget: Optional[WidgetLike] = None

    def on_click(self) -> str:
        return f"Clicked on {self._linked().name}"

    def on_hover(self) -> str:
        return f"Hovering over {self._linked().name}"

    def _linked(self) -> WidgetLike:
        if self.widget is None:
            raise RuntimeError("Interactive twin is not linked to a widget")
        return self.widget


class Widget:
    """The composite twin class that links Graphic and Interactive."""

    def __init__(self, name: str, x: int = 0, y: int = 0) -> None:
        self.name = name
        self.x = x
        self.y = y
        self.width = 100
        self.height = 50

        self._graphic = Graphic()
        self._graphic.widget = self
        self._interactive = Interactive()
        self._interactive.widget = self

    def draw(self) -> str:
        return self._graphic.draw()

    def resize(self, width: int, height: int) -> str:
        return self._graphic.resize(width, height)

    def on_click(self) -> str:
        return self._interactive.on_click()

    def on_hover(self) -> str:
        return self._interactive.on_hover()

    def get_graphic(self) -> Graphic:
        return self._graphic

    def get_interactive(self) -> Interactive:
        return self._interactive


class _FakeWidget:
    """Stub used to test a twin without a real Widget."""

    name = "Fake"
    x = 0
    y = 0
    width = 0
    height = 0


class TestTwinPattern(unittest.TestCase):
    def test_widget_delegates_to_twins(self) -> None:
        w = Widget("SubmitButton", 10, 20)
        self.assertEqual(w.draw(), "Drawing SubmitButton at (10, 20)")
        self.assertEqual(w.on_click(), "Clicked on SubmitButton")
        w.resize(200, 60)
        self.assertEqual((w.width, w.height), (200, 60))

    def test_twin_is_testable_with_a_stub(self) -> None:
        g = Graphic()
        g.widget = _FakeWidget()
        self.assertEqual(g.draw(), "Drawing Fake at (0, 0)")

    def test_unlinked_twin_fails_fast(self) -> None:
        with self.assertRaises(RuntimeError):
            Graphic().draw()

    def test_twin_can_be_swapped(self) -> None:
        w = Widget("Btn")
        custom = Graphic()
        custom.widget = w
        w._graphic = custom
        self.assertIs(w.get_graphic(), custom)


if __name__ == "__main__":
    button = Widget("SubmitButton", 10, 20)
    print(button.draw())
    print(button.on_click())
    print(button.on_hover())
    print(button.resize(200, 60))
